### PR TITLE
Revert "Revert "[core] Async submitting actor registerring (#18009)" (#18719)"

### PR DESCRIPTION
**Pull request by fishbone**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -1955,16 +1984,27 @@ Status CoreWorker::KillActor(const ActorID &actor_id, bool force_kill, bool no_r
...
-  if (!actor_manager_->CheckActorHandleExists(actor_id)) {
-    std::stringstream stream;
-    stream << "Failed to find a corresponding actor handle for " << actor_id;
-    return Status::Invalid(stream.str());
-  }
-
-  RAY_CHECK_OK(
-      gcs_client_->Actors().AsyncKillActor(actor_id, force_kill, no_restart, nullptr));
-  return Status::OK();
+  std::promise<Status> p;
```
Comment (iycheng) [inline src/ray/core_worker/core_worker.cc:1987]: This is the new code for the fixing.

Comment (ericl) [reply src/ray/core_worker/core_worker.cc:1987]: What was the root cause of the issue? Sounds like thread-safety?

Comment (iycheng) [reply src/ray/core_worker/core_worker.cc:1987]: No. It's just a race condition. It's trying to kill the actor that's pending registration.

Comment (ericl) [reply src/ray/core_worker/core_worker.cc:1987]: Got it.

**Review by ericl — APPROVED** [review]

**Pull request merged**